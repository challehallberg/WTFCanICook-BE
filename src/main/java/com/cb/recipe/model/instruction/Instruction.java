package com.cb.recipe.model.instruction;

import lombok.Builder;
import lombok.Data;

import java.util.List;

@Data
@Builder
public class Instruction {
    // The name of this instruction set (can be empty)
    //@Nullable
    private String name;

    // The steps associated with this instruction set
    private List<InstructionStep> steps;
}
