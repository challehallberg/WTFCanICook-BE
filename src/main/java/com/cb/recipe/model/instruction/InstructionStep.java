package com.cb.recipe.model.instruction;

import lombok.Builder;
import lombok.Data;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Optional;

@Data
@Builder
public class InstructionStep {
    // The order number of this step
    private int number;

    // The textual description of the step
    private String step;

    // Optional: the duration for this step (if provided)
    @Nullable
    private StepLength length;

    // A list of equipment names used in this step
    @Nullable
    private List<Equipment> equipment;

    // A list of ingredient names used in this step
    @Nullable
    private List<String> ingredients;

    public Optional<StepLength> getOptionalLength() {
        return Optional.ofNullable(length);
    }

    // Optional getter for equipment
    public Optional<List<Equipment>> getOptionalEquipment() {
        return Optional.ofNullable(equipment);
    }

    // Optional getter for ingredients
    public Optional<List<String>> getOptionalIngredients() {
        return Optional.ofNullable(ingredients);
    }
}