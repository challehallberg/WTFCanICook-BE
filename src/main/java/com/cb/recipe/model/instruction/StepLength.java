package com.cb.recipe.model.instruction;

import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class StepLength {
    // The duration number (e.g., 15)
    private double number;

    // The duration unit (e.g., "minutes")
    private String unit;
}