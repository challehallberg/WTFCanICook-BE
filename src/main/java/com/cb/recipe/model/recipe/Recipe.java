package com.cb.recipe.model.recipe;

import lombok.Builder;
import lombok.Data;

import java.math.BigDecimal;
import java.util.List;

@Data
@Builder
public class Recipe {
    private int id;
    private String title;
    private String image;
    private String imageType;
    private int likes;

    private int missedIngredientCount;
    private List<Ingredient> missedIngredients;

    private BigDecimal usedIngredientCount;
    private List<Ingredient> usedIngredients;

    private int readyInMinutes;
    private boolean vegan;
    private boolean vegetarian;
    private boolean glutenFree;
    private int servings;
}
