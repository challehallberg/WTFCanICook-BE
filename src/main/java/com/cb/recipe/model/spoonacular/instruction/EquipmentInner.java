package com.cb.recipe.model.spoonacular.instruction;
import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class EquipmentInner {
    /**
     * The equipment id.
     */
    private int id;

    /**
     * The equipment image.
     */
    private String image;

    /**
     * The name of the equipment.
     */
    private String name;

    /**
     * Optional temperature information for the equipment.
     */
    private TemperatureInner temperature;
}
