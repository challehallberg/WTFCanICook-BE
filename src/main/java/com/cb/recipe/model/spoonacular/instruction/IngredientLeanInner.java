package com.cb.recipe.model.spoonacular.instruction;
import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class IngredientLeanInner {
    /**
     * The ingredient id.
     */
    private int id;

    /**
     * The ingredient image.
     */
    private String image;

    /**
     * The name of the ingredient.
     */
    private String name;
}
