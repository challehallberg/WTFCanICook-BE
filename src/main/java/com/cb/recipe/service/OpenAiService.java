package com.cb.recipe.service;

import com.cb.recipe.client.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.Base64;
import java.util.concurrent.TimeUnit;

@Service
@Slf4j
public class OpenAiService {
    private final OpenAiClient client;
    private final Cache<String, String> imageResponseCache = Caffeine.newBuilder()
            .expireAfterWrite(10, TimeUnit.MINUTES)
            .build();

    public OpenAiService(final OpenAiClient client) {
        this.client = client;
    }

    public String processImage(final MultipartFile file) {
        final String fileName = file.getOriginalFilename();
        final String cachedIngredients = imageResponseCache.getIfPresent(fileName);

        if (cachedIngredients != null) {
            log.info("File already cached, retrieving cached ingredients");
            return cachedIngredients;
        }
        try {
            log.info("Processing image");
            final String response = client.processImage(getRequestBody(encodeImageToBase64(file)));

            final String ingredients = extractDataFromResponse(response);

            imageResponseCache.put(fileName, ingredients);

            return ingredients;
        } catch (Exception e) {
            log.warn(e.getMessage());
            return "";
        }
    }

    private String extractDataFromResponse(final String response) {
        try {
            final ObjectMapper objectMapper = new ObjectMapper();

            // Parse JSON response
            final JsonNode rootNode = objectMapper.readTree(response);

            return rootNode
                    .path("choices")     // Navigate to "choices"
                    .get(0)                 // Get the first choice
                    .path("message")     // Navigate to "message"
                    .path("content")     // Extract "content"
                    .asText();
        } catch (Exception e) {
            log.warn("Could not extract data from response");
            return "";
        }
    }

    private String getRequestBody(final String base64Image) {
        return """
                {
                    "model": "gpt-4o-mini",
                    "messages": [
                        {
                            "role": "user",
                            "content": [
                                {
                                    "type": "text",
                                    "text": "Analyze the attached image and identify all food ingredients present in it. List them in the format: ingredient1, ingredient2, ... If there are none, return an empty String"
                                },
                                {
                                    "type": "image_url",
                                    "image_url": {
                                        "url": "data:image/jpeg;base64,%s"
                                    }
                                }
                            ]
                        }
                    ]
                }
                """.formatted(base64Image);
    }

    private String encodeImageToBase64(final MultipartFile image) throws IOException {
        final byte[] fileBytes = image.getBytes();
        return Base64.getEncoder().encodeToString(fileBytes);
    }
}
