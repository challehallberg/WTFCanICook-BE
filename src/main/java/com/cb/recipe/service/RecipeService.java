package com.cb.recipe.service;

import com.cb.recipe.client.SpoonacularClient;
import com.cb.recipe.mapper.InstructionMapper;
import com.cb.recipe.mapper.RecipeMapper;
import com.cb.recipe.model.instruction.Instruction;
import com.cb.recipe.model.recipe.Recipe;
import com.cb.recipe.model.spoonacular.instruction.InstructionInner;
import com.cb.recipe.model.spoonacular.recipe.RecipeInner;
import com.cb.recipe.model.spoonacular.recipeInformation.RecipeInformationInner;
import com.github.benmanes.caffeine.cache.Cache;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.List;

@Service
@Slf4j
public class RecipeService {
    private final SpoonacularClient client;
    private final Cache<String, List<Recipe>> recipeCache;
    private final Cache<Integer, List<Instruction>> instructionsCache;

    public RecipeService(final SpoonacularClient client,
                         @Qualifier("recipeCache") final Cache<String, List<Recipe>> recipeCache,
                         @Qualifier("instructionsCache") final Cache<Integer, List<Instruction>> instructionsCache) {
        this.client = client;
        this.recipeCache = recipeCache;
        this.instructionsCache = instructionsCache;
    }

    public List<Recipe> getRecipe(final String ingredients) {
        final List<Recipe> cachedRecipes = recipeCache.getIfPresent(ingredients);
        if (cachedRecipes != null) {
            log.info("Recipes based on these ingredients are already cached, retrieving cached recipes");
            return cachedRecipes;
        }

        final int number = 10;                 // Integer | The maximum number of items to return (between 1 and 100). Defaults to 10.
        final int ranking = 1;                 // Integer | Whether to maximize used ingredients (1) or minimize missing ingredients (2) first.
        final boolean ignorePantry = true;     // Boolean | Whether to ignore typical pantry items, such as water, salt, flour, etc.

        try {
            final List<RecipeInner> recipesByIngredients = client.getRecipesByIngredients(ingredients, number, ranking, ignorePantry);

            final List<Recipe> recipes = RecipeMapper.map(recipesByIngredients);

            recipeCache.put(ingredients, recipes);

            return recipes;
        } catch (Exception e) {
            log.error(e.getMessage());
            return Collections.emptyList();
        }
    }

    public List<RecipeInformationInner> getRecipeInformationByIds(final String recipeIds) {
        return client.getRecipeInformationBulk(recipeIds, false);
    }

    public List<Instruction> getAnalyzedRecipeInstructions(final int recipeId) {
        final List<Instruction> cachedInstructions = instructionsCache.getIfPresent(recipeId);
        if (cachedInstructions != null) {
            log.info("Instructions for {} are already cached, retrieving cached instructions", recipeId);
            return cachedInstructions;
        }

        final List<InstructionInner> analyzedInstructions = client.getAnalyzedInstructions(recipeId, false);
        final List<Instruction> mappedInstructions = InstructionMapper.map(analyzedInstructions);

        instructionsCache.put(recipeId, mappedInstructions);
        return mappedInstructions;
    }
}
