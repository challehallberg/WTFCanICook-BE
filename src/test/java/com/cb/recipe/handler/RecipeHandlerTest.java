package com.cb.recipe.handler;

import com.cb.recipe.faker.instruction.InstructionFaker;
import com.cb.recipe.faker.recipe.RecipeFaker;
import com.cb.recipe.faker.recipe.RecipeInformationInnerFaker;
import com.cb.recipe.model.instruction.Instruction;
import com.cb.recipe.model.recipe.Recipe;
import com.cb.recipe.service.OpenAiService;
import com.cb.recipe.service.RecipeService;
import com.cb.recipe.util.UnitTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.web.multipart.MultipartFile;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class RecipeHandlerTest extends UnitTest {

    @Mock
    private OpenAiService openAiService;

    @Mock
    private RecipeService recipeService;

    @Mock
    private MultipartFile file1;

    @Mock
    private MultipartFile file2;

    @InjectMocks
    private RecipeHandler recipeHandler;

    @Test
    void givenValidFilesWithNoBytes_whenGetRecipes_thenExpectEmptyList() {
        when(openAiService.processImage(file1))
                .thenReturn("");
        when(openAiService.processImage(file2))
                .thenReturn("");

        final List<Recipe> result = recipeHandler.getRecipes(file1, file2);

        assertEquals(Collections.emptyList(), result);
        verify(recipeService, never()).getRecipe(anyString());
    }

    @Test
    void givenValidFile_whenGetRecipes_thenExpectRecipeList() {
        final int recipeId = 123;

        when(openAiService.processImage(file1))
                .thenReturn("apple, banana");

        when(recipeService.getRecipe("apple, banana"))
                .thenReturn(Arrays.asList(RecipeFaker.recipe()
                        .withId(recipeId)
                        //Enrichment fields are not set initially.
                        .withReadyInMinutes(0)
                        .withVegan(false)
                        .withVegetarian(false)
                        .withGlutenFree(false)
                        .withServings(0)
                        .create()));

        when(recipeService.getRecipeInformationByIds(Mockito.anyString()))
                .thenReturn(Arrays.asList(RecipeInformationInnerFaker.recipeInformation()
                        .withId(recipeId)
                        .create()));

        final List<Recipe> result = recipeHandler.getRecipes(file1);

        snapshot(result);
    }

    @Test
    void givenTwoValidFiles_whenGetRecipes_thenExpectRecipeList() {
        final int recipeId1 = 123;
        final int recipeId2 = 321;

        when(openAiService.processImage(file1))
                .thenReturn("apple, banana");

        when(recipeService.getRecipe("apple, banana"))
                .thenReturn(Arrays.asList(RecipeFaker.recipe()
                        .withId(recipeId1)
                        //Enrichment fields are not set initially.
                        .withReadyInMinutes(0)
                        .withVegan(false)
                        .withVegetarian(false)
                        .withGlutenFree(false)
                        .withServings(0)
                        .create(), RecipeFaker.recipe(new Random(2))
                        .withId(recipeId2)
                        //Enrichment fields are not set initially.
                        .withReadyInMinutes(0)
                        .withVegan(false)
                        .withVegetarian(false)
                        .withGlutenFree(false)
                        .withServings(0)
                        .create()));

        when(recipeService.getRecipeInformationByIds(Mockito.anyString()))
                .thenReturn(Arrays.asList(RecipeInformationInnerFaker.recipeInformation()
                        .withId(recipeId1)
                        .create(), RecipeInformationInnerFaker.recipeInformation(new Random(2))
                        .withId(recipeId2)
                        .create()));

        final List<Recipe> result = recipeHandler.getRecipes(file1);

        snapshot(result);
    }

    @Test
    void givenValidRecipeId_whenGetAnalyzedRecipeInstructions_thenExpectedInstructions() {
        final int recipeId = 101;

        when(recipeService.getAnalyzedRecipeInstructions(recipeId))
                .thenReturn(Arrays.asList(InstructionFaker.instruction().create()));

        List<Instruction> result = recipeHandler.getInstructions(recipeId);

        snapshot(result);

        verify(recipeService, times(1)).getAnalyzedRecipeInstructions(recipeId);
    }
}
