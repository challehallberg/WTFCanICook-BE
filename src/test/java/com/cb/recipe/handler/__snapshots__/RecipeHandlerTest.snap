com.cb.recipe.handler.RecipeHandlerTest.givenTwoValidFiles_whenGetRecipes_thenExpectRecipeList=[
  [
    {
      "id": 123,
      "title": "Recipe 985",
      "image": "https://example.com/image985.jpg",
      "imageType": "jpg",
      "likes": 88,
      "missedIngredientCount": 2,
      "missedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        }
      ],
      "usedIngredientCount": 3,
      "usedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        },
        {
          "id": 108,
          "name": "Ingredient 108",
          "amount": 3.6403890715990754,
          "unit": "cup",
          "image": "https://example.com/ingredient108.jpg"
        }
      ],
      "readyInMinutes": 48,
      "vegan": false,
      "vegetarian": false,
      "glutenFree": false,
      "servings": 5
    },
    {
      "id": 321,
      "title": "Recipe 108",
      "image": "https://example.com/image108.jpg",
      "imageType": "jpg",
      "likes": 72,
      "missedIngredientCount": 0,
      "missedIngredients": [ ],
      "usedIngredientCount": 2,
      "usedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        }
      ],
      "readyInMinutes": 32,
      "vegan": true,
      "vegetarian": false,
      "glutenFree": false,
      "servings": 1
    }
  ]
]


com.cb.recipe.handler.RecipeHandlerTest.givenValidFile_whenGetRecipes_thenExpectRecipeList=[
  [
    {
      "id": 123,
      "title": "Recipe 985",
      "image": "https://example.com/image985.jpg",
      "imageType": "jpg",
      "likes": 88,
      "missedIngredientCount": 2,
      "missedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        }
      ],
      "usedIngredientCount": 3,
      "usedIngredients": [
        {
          "id": 360,
          "name": "Ingredient 360",
          "amount": 8.48296889908355,
          "unit": "cup",
          "image": "https://example.com/ingredient360.jpg"
        },
        {
          "id": 985,
          "name": "Ingredient 985",
          "amount": 1.9042589469362396,
          "unit": "g",
          "image": "https://example.com/ingredient985.jpg"
        },
        {
          "id": 108,
          "name": "Ingredient 108",
          "amount": 3.6403890715990754,
          "unit": "cup",
          "image": "https://example.com/ingredient108.jpg"
        }
      ],
      "readyInMinutes": 48,
      "vegan": false,
      "vegetarian": false,
      "glutenFree": false,
      "servings": 5
    }
  ]
]


com.cb.recipe.handler.RecipeHandlerTest.givenValidRecipeId_whenGetAnalyzedRecipeInstructions_thenExpectedInstructions=[
  [
    {
      "name": "Instruction Set 88",
      "steps": [
        {
          "number": 1,
          "step": "Step description 4",
          "equipment": [ ],
          "ingredients": [ ]
        },
        {
          "number": 2,
          "step": "Step description 7",
          "length": {
            "number": 21.611449199578143,
            "unit": "minutes"
          },
          "equipment": [ ],
          "ingredients": [
            "Ingredient 64",
            "Ingredient 63"
          ]
        },
        {
          "number": 3,
          "step": "Step description 5",
          "length": {
            "number": 27.41756394766841,
            "unit": "minutes"
          },
          "equipment": [ ],
          "ingredients": [ ]
        },
        {
          "number": 4,
          "step": "Step description 3",
          "equipment": [ ],
          "ingredients": [ ]
        },
        {
          "number": 5,
          "step": "Step description 10",
          "length": {
            "number": 26.069426661179147,
            "unit": "minutes"
          },
          "equipment": [
            {
              "name": "Equipment 5",
              "temperature": {
                "number": 215.989270869342,
                "unit": "Fahrenheit"
              }
            }
          ],
          "ingredients": [ ]
        },
        {
          "number": 6,
          "step": "Step description 6",
          "equipment": [
            {
              "name": "Equipment 75",
              "temperature": {
                "number": 219.49479796024917,
                "unit": "Fahrenheit"
              }
            }
          ],
          "ingredients": [
            "Ingredient 38",
            "Ingredient 78"
          ]
        }
      ]
    }
  ]
]